source/common/router/config_impl.h

```diff
@@ -374,6 +374,26 @@ class PathRouteEntryImpl : public RouteEntryImplBase {
 };
 
 /**
+ * Route entry implementation for regular expression match routing.
+ */
+class RegexRouteEntryImpl : public RouteEntryImplBase {
+public:
+  RegexRouteEntryImpl(const VirtualHostImpl& vhost, const envoy::api::v2::Route& route,
+                      Runtime::Loader& loader);
+
+  // Router::RouteEntry
+  void finalizeRequestHeaders(Http::HeaderMap& headers) const override;
+
+  // Router::Matchable
+  RouteConstSharedPtr matches(const Http::HeaderMap& headers, uint64_t random_value) const override;
+
+private:
+  const char* findQueryStringStart(const Http::HeaderString& path) const;
```

Comment:
Nit: can be `static` or internal to `config_impl.cc` in the anonymous namespace (alternatively, elevate to somewhere more central as a general HTTP helper utility, https://github.com/envoyproxy/envoy/blob/master/source/common/http/utility.h#L42 is a good location).